packages/core/src/utils/logger.test.ts

```diff
@@ -0,0 +1,133 @@
+import { CreateLog } from '@logto/schemas';
+
+import { OmitAutoSetFields } from '@/database/utils';
+import { initTokenExchangeSuccessLogger } from '@/utils/logger';
+import { createContextWithRouteParameters } from '@/utils/test-utils';
+
+const insertLog: (data: OmitAutoSetFields<CreateLog>) => Promise<void> = jest.fn();
+
+jest.mock('@/queries/log', () => ({
+  insertLog: jest.fn(async (data) => insertLog(data)),
+}));
+
+describe('initTokenExchangeSuccessLogger', () => {
+  const ip = '192.168.0.1';
+  const userAgent =
+    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.51 Safari/537.36';
+  const request = { ip, headers: { 'user-agent': userAgent } };
+
+  const userId = 'userId';
+  const entities = { Account: { accountId: userId } };
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('should log type CodeExchangeToken when grant type is authorization_code', async () => {
+    const parameters = { grant_type: 'authorization_code', code: 'codeValue' };
+    const ctx = {
+      ...createContextWithRouteParameters({ headers: { 'user-agent': userAgent } }),
+      request,
+      oidc: { entities, params: parameters },
+      body: {
+        access_token: 'newAccessTokenValue',
+        refresh_token: 'newRefreshTokenValue',
+        id_token: 'newIdToken',
+        scope: 'openid offline-access',
+      },
+    };
+    ctx.request.ip = ip;
+
+    const logger = initTokenExchangeSuccessLogger();
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-expect-error
+    await logger(ctx);
+    expect(insertLog).toHaveBeenCalledWith(
+      expect.objectContaining({
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+        id: expect.any(String),
+        type: 'CodeExchangeToken',
+        payload: {
+          ip,
+          issued: ['accessToken', 'refreshToken', 'idToken'],
+          params: parameters,
+          result: 'Success',
+          scope: 'openid offline-access',
+          userAgent,
+          userId,
+        },
+      })
+    );
+  });
+
+  it('should log type RefreshTokenExchangeToken when grant type is refresh_code', async () => {
+    const parameters = { grant_type: 'refresh_token', refresh_token: 'refreshTokenValue' };
+    const ctx = {
+      ...createContextWithRouteParameters({ headers: { 'user-agent': userAgent } }),
+      request,
+      oidc: { entities, params: parameters },
+      body: {
+        access_token: 'newAccessTokenValue',
+        refresh_token: 'newRefreshTokenValue',
+        id_token: 'newIdToken',
+        scope: 'openid offline-access',
+      },
+    };
+
+    const logger = initTokenExchangeSuccessLogger();
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-expect-error
```

Comment:
what error do we expect?

Reply:
just ignore tyep check to mock `ctx`
![image](https://user-images.githubusercontent.com/10594507/168218527-52e9029e-1710-4ec8-b4f7-6597e6127136.png)

if use `@ts-ignore`, it will be auto replaced with `@ts-expect-error` by lint.
![image](https://user-images.githubusercontent.com/10594507/168218623-c8778b70-5e72-4a07-823b-44244bb8e0bb.png)

any better idea?

Reply:
No better idea. 😂 I'm good with it for now